frontend/javascripts/admin/api/token.ts

```diff
@@ -33,22 +35,48 @@ export function getSharingTokenFromUrlParameters(): string | null | undefined {
   return null;
 }
 
-export function doWithToken<T>(fn: (token: string) => Promise<T>, tries: number = 1): Promise<any> {
-  const sharingToken = getSharingTokenFromUrlParameters();
+function removeErrorToast(error: any) {
+  if ("errors" in error && Array.isArray(error.errors)) {
+    error.errors.forEach((errorText: string) => {
+      if (errorText.includes("Token may be expired")) {
+        Toast.close(errorText);
+      }
+    });
+  }
+}
+
+export async function doWithToken<T>(
+  fn: (token: string) => Promise<T>,
+  tries: number = 1,
+  useURLTokenIfAvailable: boolean = true,
+): Promise<any> {
+  let token =
+    useURLTokenIfAvailable && shouldUseURLToken ? getSharingTokenFromUrlParameters() : null;
 
-  if (sharingToken != null) {
-    return fn(sharingToken);
+  if (token == null) {
+    tokenPromise = tokenPromise == null ? requestUserToken() : tokenPromise;
+  } else {
+    tokenPromise = Promise.resolve(token);
   }
 
-  if (!tokenPromise) tokenPromise = requestUserToken();
-  return tokenPromise.then(fn).catch((error) => {
+  return tokenPromise.then(fn).catch(async (error) => {
     if (error.status === 403) {
       console.warn("Token expired. Requesting new token...");
       tokenPromise = requestUserToken();
 
       // If three new tokens did not fix the 403, abort, otherwise we'll get into an endless loop here
       if (tries < 3) {
-        return doWithToken(fn, tries + 1);
+        // If using the url sharing token failed, we try the user specific token instead.
+        const result = await doWithToken(fn, tries + 1, false);
+        // Upon successful retry with own token, discard the url token.
+        if (useURLTokenIfAvailable) {
+          shouldUseURLToken = false;
+          Toast.info(
```

Comment:
I think this is a technical detail that we shouldn't bother users with. I would replace the `Toast` with a `console.log()` for debugging purposes but not actively show it in the UI.

Reply:
So just hiding the error toast is enough you think? My thought was to somewhat inform the user that saving / or some other request worked on second try.

Therefore, maybe a toast with a message like "Request succeeded on upon retry" or something similar?

Reply:
But yeah, I agree that the technical details are definitely too much in my toast message